Fix & test Windows bracketed pasting issue

Splines opened the pull request:
Closes #18.

This fixes an annoying Windows bug. See [the Wiki](https://github.com/Manim-Notebook/manim-notebook/wiki/%F0%9F%A4%A2-Troubleshooting#windows-paste-patch) for more details. What this PR does is apply the [Python patch I presented here](https://github.com/prompt-toolkit/python-prompt-toolkit/issues/1894#issuecomment-2601043935). It will create the respective patch files under your `site-packages` of your Python installation (it also works in virtual environments). Two new files are added, see the Wiki.

TODO:
- [x] Remove print statements in patch

Splines reviewed (commented):
Hey @mitkonikov and @RickLuiken 👋 This PR is now ready for review and hopefully applies a final fix for the Windows bug #18. Your local installation of Python will be automatically patched when you start the extension. All other fixes we've tried so far are removed in this PR.

- Everything should work out of the box. Please test again with the laggy scene and also increase the grid size to make it really laggy. It should still work at all times.
- Additionally, please test that this fix still works after you have typed `reload(<line_number>)` inside the ManimShell. This will reload all user-defined modules. The patch should be unaffected, but you never know, so please test this as well. Thanks!

In the pipeline, everything works now, see the latest run (for the preview, we do test that the full animation has run through).

Splines commented:
@mitkonikov, @RickLuiken : Just to let you know, I will merge this PR very soon to move on. The patch works flawlessly in our pipeline on a Windows server, so I hope that it'll work out for you locally as well on a Windows PC. Feel free to open #18 again, should this not be the case.

mitkonikov reviewed (changes requested):
@Splines On my machine the patch command fails, but I've given all of the necessary comments to fix this issue. When you fix this as per the review, just tag me once again, I'll try it out again. The main thing is not to open a terminal to start a process, but instead just run the `exec` command. This bypasses the terminal, its unreliability, the fact that the user can have **different types** of terminals that support different things and any wait time associated with the terminal. All the info is in the comments.

Splines commented on src/extension.ts:
```diff
@@ -49,23 +50,31 @@ export async function activate(context: vscode.ExtensionContext) {
-    manimglPath = pythonEnvToManimglPath(manimglPath);
+
+  if (process.platform === "win32") {
+    const python3Path = pythonEnvPath
+      ? getBinaryPathInPythonEnv(pythonEnvPath, "python3")
+      : "python3";
```
So using just hardcoded `python` instead of `python3` should work? I will change that, thanks for pointing it out.

Splines replied:
Done as of ddf4890d30

Splines commented on src/patches/applyPatches.ts:
```diff
@@ -0,0 +1,102 @@
+ * applied, and false otherwise. Might never resolve, so the caller should
+ * let this promise race with a timeout promise.
+ */
+async function lookForPatchSuccessfullyAppliedMessage(
+  terminal: Terminal, command: string): Promise<boolean> {
+  return new Promise<boolean>(async (resolve, _reject) => {
+    onTerminalOutput(terminal, (data: string) => {
+      const success = data.includes("Manim Notebook: Windows paste patch applied");
+      if (!success) {
+        return;
+      }
+      terminal.dispose();
+      resolve(true);
+    });
+
+    window.onDidEndTerminalShellExecution((event) => {
+      if (event.terminal !== terminal) {
+        return;
+      }
+      if (event.exitCode !== 0) {
+        Logger.error("Windows Paste Patch Terminal: shell exited with error code");
+        resolve(false);
+      }
+    });
```
> Is there a reason as to why are you opening a terminal 

My original intention with this instead of exec was to give the user a chance to see what is wrong if the patch fails. But I see that opening a new terminal is not the most reliable thing, so will use `exec` instead.

Splines replied:
Done as of fbc7eacee93

Splines commented on src/patches/applyPatches.ts:
```diff
@@ -0,0 +1,102 @@
+  terminal: Terminal, command: string): Promise<boolean> {
+  return new Promise<boolean>(async (resolve, _reject) => {
+    onTerminalOutput(terminal, (data: string) => {
+      const success = data.includes("Manim Notebook: Windows paste patch applied");
```
Ups, true ;)

Splines replied:
Done as of 958088cf

Splines commented:
@mitkonikov Thanks a lot for your review. I've implemented your requested changes and the pipeline still runs through. So feel free to try it out again. (Note that I've also removed the 🐒 print statements from the Python patch).

Splines commented:
@mitkonikov I've also added a new test to assert that the patch is successfully applied on Windows, however it fails. For the check, I spy the `Logger.info` call and assert it to be called with the string `Windows paste patch successfully applied`.

Maybe you could try out the test locally and find out why this doesn't work? E.g. [see this line in the Actions run](https://github.com/Manim-Notebook/manim-notebook/actions/runs/13029567177/job/36345630766?pr=117#step:10:62):

```
AssertError: expected info to have been called at least once but was never called
```

Splines commented:
Ok, so if I comment out the `if (process.platform !== "win32") {` checks locally, the Windows paste patch test runs through and successfully creates the respective patch files inside the project's `tmp/manimVenv/lib/python3.<version>/site-packages/` folder. If I change the `SUCCESS_SIGNATURE` in the Python patch file, the test fails for me locally, indicating that I really do test what I wanted to test.

Despite this, it does not work in GitHub Actions pipeline, so I'm curious to know if it works locally for you, @mitkonikov

mitkonikov commented on src/patches/install_windows_paste_patch.py:
```diff
@@ -0,0 +1,95 @@
+# Construct a string that we check against, without the string being present
+# in the source code itself.
+SUCCESS_SIGNATURE = "42" + "0" * 4 + "43" + "Manim" + "Notebook" + "31415"
+print("🐒 Manim Notebook: Windows paste patch applied")
```
Remove the emoji from here, because when you run exec, it converts it to some other encoding and afterwards the python -c command fails because of it. I have to research why exactly this happen, but since it's not a priority, just remove it.

Splines replied:
Done as of 3fcf20f39bad6

mitkonikov commented on src/patches/applyPatches.ts:
```diff
@@ -0,0 +1,93 @@
+  patch = patch.replace(/'/g, "\\'");
+  patch = patch.replace(/\n/g, "\\n");
+  patch = patch.replace(/\r/g, "\\r");
+  const patchCommand = `${pythonBinary} -c "exec('''${patch}''')"`;
```
For some reason, I get the following `pythonBinary` here:

```sh
C:\\GitHub\\animations\\.venv\\Scripts\\python.exe\\Scripts\\python
```

Splines replied:
Done as of 6d60b78799c4

mitkonikov reviewed (changes requested):
I'm writing this review here quickly because my laptop battery will die shortly. Here are some things I've uncovered. With these things fixed, I get the command to be successfully executed. But I have to try it again after these are fixed. I don't have the resources now to check the get python path function, but If you give me couple of hours I can report more details (if you don't find out by then). Please tag me again and request a review. Also try to work to determine the Manim version by running the exec command, since it's way more reliable than doing a terminal proxy. The exec command gave me a lot of details in `stderr`, so you can check that if you have any problems.

Splines commented:
> Also try to work to determine the Manim version by running the exec command, since it's way more reliable than doing a terminal proxy. 

Yes, will do that when I tackle #119 soon in another PR.

> The exec command gave me a lot of details in `stderr`

I already assert that `Logger.error()` is never called in the CI/CD (`Logger.error()` is called with `stderr` should there be any). However, in the pipeline, the "Applies Windows paste patch" test fails [due to timeout](https://github.com/Manim-Notebook/manim-notebook/actions/runs/13035809761/job/36366155144?pr=117#step:10:66) and until then, `Logger.error()` was never called, see [the test code](https://github.com/Manim-Notebook/manim-notebook/blob/6d60b78799c4048be3b8a7d4ee7dcdea3b52ac69/tests/activation.test.ts#L53-L81). Note that I've already tried to increase the timeout to 10s but that doesn't help (and we do let the patch fail already after 4s).

mitkonikov commented on src/patches/install_windows_paste_patch.py:
```diff
@@ -0,0 +1,95 @@
+print("Manim Notebook: Windows paste patch applied")
+print(SUCCESS_SIGNATURE)
+print(
+    "Fore more information, see https://github.com/Manim-Notebook/manim-notebook/wiki/%F0%9F%A4%A2-Troubleshooting#windows-paste-patch"
```
Fix typo `For more`.

Splines replied:
thx, fixed in abb68397

mitkonikov reviewed (approved):
Cool! For me, it works now! It applies the patch and there's no problem executing the laggy blocks. The only thing that remains bugging me is the version check. But you can fix that in another PR. I will now take a look at why the Windows pipeline on CI/CD fails. Is there a way to look at the logs from the pipeline? Can you make the logger print on STDOUT when in the pipeline? So, we can see all of the logs instead of just guessing what's wrong?

Splines commented:
> Can you make the logger print on STDOUT when in the pipeline?

Done in the last commit. You can see the output [here](https://github.com/Manim-Notebook/manim-notebook/actions/runs/13040807769/job/36382010838?pr=117#step:10:77). Usually, you can open the pipeline logs by clicking on the dot or arrow next to the commit hash on GitHub, then on `Details`.

<img src="https://github.com/user-attachments/assets/29047963-2dc5-4757-a3c0-4e3e3e194ed6" width=450>

(If however by "log" you mean the Manim Notebook Log, no that's unfortunately not possible. But we can of course stub the Logger should those messages be important in the pipeline.)

Don't know why I didn't have the idea to just `console.log()` earlier 🙈 But I did stub the `Logger.error`, however sinon.js never triggered it, weird... But at least, I see what is going on now:

```
💨 Windows Paste Patch Exec.
Manim Notebook: Windows paste patch applied
42000043ManimNotebook31415
For more information, see https:\github.com\Manim-Notebook\manim-notebook\wiki\%F0%9F%A4%A2-Troubleshooting#windows-paste-patch

<string>:1: SyntaxWarning: invalid escape sequence '\g'
<string>:7: SyntaxWarning: invalid escape sequence '\g'
<string>:94: SyntaxWarning: invalid escape sequence '\g'

💨 -----
```

@mitkonikov Could you also check that the tests are running through locally for you?

Splines commented:
@mitkonikov At the end, I was just fed up with encoding errors, so I've decided to base64-encode the file and then decode it in Python itself. And finally, it works 💫

Could you please remove the patch files manually in your Python installation that you use when running the extension, as well as inside `/tmp/manimVenv/<subfolder>`, then:
- Run the tests
- Run the extension and test manually via the laggy scene

Hopefully, this will be the last time ;) In the latest commit, I've removed the temporary `console.log()` statements and ran all tests. And the pipeline is green now ✅

mitkonikov reviewed (approved):
LGTM! The Windows tests work locally too! So, to me this is already mergable! And we can move on to testing the Manim `--version` thing, because this bugs me out and I had to bypass it locally to run the tests properly.

mitkonikov commented:
@Splines You are welcomed, it was a pleasure! Ideally, we should contact Grant and maybe try to one day fix it inside Manim itself (maybe by adding a flag or something, is this possible from inside Manim?). Because this is still a very invasive patch. But, for now, it's all good. ✨

Splines commented:
> we should contact Grant and maybe try to one day fix it inside Manim itself 

Your motivation to do so is obviously stronger since you're a Windows users ;) Feel free to open a new issue [on the Manim repo](https://github.com/3b1b/manim).

> is this possible from inside Manim?

If we can fix it here in TypeScript, it can also be fixed directly in Manim, yes ;) But in the best case, the [underlying issue](https://github.com/prompt-toolkit/python-prompt-toolkit/issues/1894) is fixed instead of deploying patches everywhere. But who knows when this will be the case...
